File: packages/api-client/src/types/report.ts
import { UUID } from './common';

export interface DashboardSummary {
  today: TodaySummary;
  this_month: MonthSummary;
  outstanding: OutstandingSummary;
  cash_position: CashPositionSummary;
  recent_transactions: TransactionSummary[];
  overdue_invoices: InvoiceSummary[];
}

export interface TodaySummary {
  sales: number;
  expenses: number;
  net: number;
  invoices_created: number;
  payments_received: number;
}

export interface MonthSummary {
  sales: number;
  expenses: number;
  net: number;
  sales_change_percent: number;
}

export interface OutstandingSummary {
  receivables: number;
  payables: number;
}

export interface CashPositionSummary {
  cash_in_hand: number;
  bank_balance: number;
  total: number;
}

export interface TransactionSummary {
  id: UUID;
  date: string;
  type: string;
  description: string;
  amount: number;
  party_name?: string;
}

export interface InvoiceSummary {
  id: UUID;
  invoice_number: string;
  customer_name: string;
  amount: number;
  due_date: string;
  days_overdue: number;
}

export interface ProfitLossReport {
  period: ReportPeriod;
  revenue: RevenueSection;
  expenses: ExpenseSection;
  gross_profit: number;
  gross_margin_percent: number;
  operating_profit: number;
  net_profit: number;
  net_margin_percent: number;
}

export interface ReportPeriod {
  from: string;
  to: string;
}

export interface RevenueSection {
  sales: number;
  other_income: number;
  total: number;
}

export interface ExpenseSection {
  cost_of_goods_sold: number;
  operating_expenses: OperatingExpenseSection;
  total: number;
}

export interface OperatingExpenseSection {
  rent: number;
  salaries: number;
  utilities: number;
  marketing: number;
  other: number;
  total: number;
}

export interface BalanceSheet {
  as_of_date: string;
  assets: AssetsSection;
  liabilities: LiabilitiesSection;
  equity: EquitySection;
}

export interface AssetsSection {
  current_assets: CurrentAssetsSection;
  fixed_assets: number;
  total_assets: number;
}

export interface CurrentAssetsSection {
  cash: number;
  bank: number;
  accounts_receivable: number;
  inventory: number;
  total: number;
}

export interface LiabilitiesSection {
  current_liabilities: CurrentLiabilitiesSection;
  total_liabilities: number;
}

export interface CurrentLiabilitiesSection {
  accounts_payable: number;
  tax_payable: number;
  total: number;
}

export interface EquitySection {
  owner_capital: number;
  retained_earnings: number;
  total_equity: number;
}

export interface GSTSummary {
  period: string;
  outward_supplies: GSTSupplies;
  inward_supplies: GSTSupplies;
  tax_liability: GSTTaxLiability;
}

export interface GSTSupplies {
  taxable_value: number;
  cgst: number;
  sgst: number;
  igst: number;
  cess: number;
  total_tax: number;
}

export interface GSTTaxLiability {
  cgst: number;
  sgst: number;
  igst: number;
  total: number;
}

export interface ReceivablesAgingReport {
  summary: AgingSummary;
  by_customer: CustomerAging[];
}

export interface AgingSummary {
  current: number;
  '1_30_days': number;
  '31_60_days': number;
  '61_90_days': number;
  over_90_days: number;
  total: number;
}

export interface CustomerAging {
  customer_id: UUID;
  customer_name: string;
  current: number;
  '1_30_days': number;
  '31_60_days': number;
  '61_90_days': number;
  over_90_days: number;
  total: number;
}

export interface CashFlowReport {
  period: ReportPeriod;
  opening_balance: number;
  operating_activities: CashFlowSection;
  investing_activities: CashFlowSection;
  financing_activities: CashFlowSection;
  net_cash_flow: number;
  closing_balance: number;
}

export interface CashFlowSection {
  inflow: number;
  outflow: number;
  net: number;
}

File: services/report-service/internal/models/report.go
package models

import (
	"time"

	"github.com/google/uuid"
)

// DashboardSummary represents the dashboard summary data
type DashboardSummary struct {
	Today      TodaySummary      `json:"today"`
	ThisMonth  MonthSummary      `json:"this_month"`
	Outstanding OutstandingSummary `json:"outstanding"`
	CashPosition CashPositionSummary `json:"cash_position"`
	RecentTransactions []TransactionSummary `json:"recent_transactions"`
	OverdueInvoices []InvoiceSummary `json:"overdue_invoices"`
}

// TodaySummary represents today's transaction summary
type TodaySummary struct {
	Sales            float64 `json:"sales"`
	Expenses         float64 `json:"expenses"`
	Net              float64 `json:"net"`
	InvoicesCreated  int     `json:"invoices_created"`
	PaymentsReceived int     `json:"payments_received"`
}

// MonthSummary represents this month's summary
type MonthSummary struct {
	Sales              float64 `json:"sales"`
	Expenses           float64 `json:"expenses"`
	Net                float64 `json:"net"`
	SalesChangePercent float64 `json:"sales_change_percent"`
}

// OutstandingSummary represents outstanding amounts
type OutstandingSummary struct {
	Receivables float64 `json:"receivables"`
	Payables    float64 `json:"payables"`
}

// CashPositionSummary represents cash position
type CashPositionSummary struct {
	CashInHand  float64 `json:"cash_in_hand"`
	BankBalance float64 `json:"bank_balance"`
	Total       float64 `json:"total"`
}

// TransactionSummary represents a transaction summary for dashboard
type TransactionSummary struct {
	ID              uuid.UUID `json:"id"`
	Date            time.Time `json:"date"`
	Type            string    `json:"type"`
	Description     string    `json:"description"`
	Amount          float64   `json:"amount"`
	PartyName       string    `json:"party_name,omitempty"`
}

// InvoiceSummary represents an invoice summary for dashboard
type InvoiceSummary struct {
	ID            uuid.UUID `json:"id"`
	InvoiceNumber string    `json:"invoice_number"`
	CustomerName  string    `json:"customer_name"`
	Amount        float64   `json:"amount"`
	DueDate       time.Time `json:"due_date"`
	DaysOverdue   int       `json:"days_overdue"`
}

// ProfitLossReport represents a P&L report
type ProfitLossReport struct {
	Period        ReportPeriod    `json:"period"`
	Revenue       RevenueSection  `json:"revenue"`
	Expenses      ExpenseSection  `json:"expenses"`
	GrossProfit   float64         `json:"gross_profit"`
	GrossMargin   float64         `json:"gross_margin_percent"`
	OperatingProfit float64       `json:"operating_profit"`
	NetProfit     float64         `json:"net_profit"`
	NetMargin     float64         `json:"net_margin_percent"`
}

// ReportPeriod represents the period for a report
type ReportPeriod struct {
	From time.Time `json:"from"`
	To   time.Time `json:"to"`
}

// RevenueSection represents revenue in P&L
type RevenueSection struct {
	Sales       float64 `json:"sales"`
	OtherIncome float64 `json:"other_income"`
	Total       float64 `json:"total"`
}

// ExpenseSection represents expenses in P&L
type ExpenseSection struct {
	CostOfGoodsSold   float64                 `json:"cost_of_goods_sold"`
	OperatingExpenses OperatingExpenseSection `json:"operating_expenses"`
	Total             float64                 `json:"total"`
}

// OperatingExpenseSection represents operating expenses
type OperatingExpenseSection struct {
	Rent      float64 `json:"rent"`
	Salaries  float64 `json:"salaries"`
	Utilities float64 `json:"utilities"`
	Marketing float64 `json:"marketing"`
	Other     float64 `json:"other"`
	Total     float64 `json:"total"`
}

// BalanceSheet represents a balance sheet report
type BalanceSheet struct {
	AsOfDate    time.Time      `json:"as_of_date"`
	Assets      AssetsSection  `json:"assets"`
	Liabilities LiabilitiesSection `json:"liabilities"`
	Equity      EquitySection  `json:"equity"`
}

// AssetsSection represents assets in balance sheet
type AssetsSection struct {
	CurrentAssets    CurrentAssetsSection `json:"current_assets"`
	FixedAssets      float64              `json:"fixed_assets"`
	TotalAssets      float64              `json:"total_assets"`
}

// CurrentAssetsSection represents current assets
type CurrentAssetsSection struct {
	Cash             float64 `json:"cash"`
	Bank             float64 `json:"bank"`
	AccountsReceivable float64 `json:"accounts_receivable"`
	Inventory        float64 `json:"inventory"`
	Total            float64 `json:"total"`
}

// LiabilitiesSection represents liabilities in balance sheet
type LiabilitiesSection struct {
	CurrentLiabilities CurrentLiabilitiesSection `json:"current_liabilities"`
	TotalLiabilities   float64                   `json:"total_liabilities"`
}

// CurrentLiabilitiesSection represents current liabilities
type CurrentLiabilitiesSection struct {
	AccountsPayable float64 `json:"accounts_payable"`
	TaxPayable      float64 `json:"tax_payable"`
	Total           float64 `json:"total"`
}

// EquitySection represents equity in balance sheet
type EquitySection struct {
	OwnerCapital     float64 `json:"owner_capital"`
	RetainedEarnings float64 `json:"retained_earnings"`
	TotalEquity      float64 `json:"total_equity"`
}

// GSTSummary represents GST summary report
type GSTSummary struct {
	Period          string            `json:"period"`
	OutwardSupplies GSTSupplies       `json:"outward_supplies"`
	InwardSupplies  GSTSupplies       `json:"inward_supplies"`
	TaxLiability    GSTTaxLiability   `json:"tax_liability"`
}

// GSTSupplies represents GST supplies (inward or outward)
type GSTSupplies struct {
	TaxableValue float64 `json:"taxable_value"`
	CGST         float64 `json:"cgst"`
	SGST         float64 `json:"sgst"`
	IGST         float64 `json:"igst"`
	Cess         float64 `json:"cess"`
	TotalTax     float64 `json:"total_tax"`
}

// GSTTaxLiability represents net tax liability
type GSTTaxLiability struct {
	CGST  float64 `json:"cgst"`
	SGST  float64 `json:"sgst"`
	IGST  float64 `json:"igst"`
	Total float64 `json:"total"`
}

// ReceivablesAgingReport represents receivables aging report
type ReceivablesAgingReport struct {
	Summary    AgingSummary       `json:"summary"`
	ByCustomer []CustomerAging    `json:"by_customer"`
}

// AgingSummary represents aging summary
type AgingSummary struct {
	Current     float64 `json:"current"`
	Days1To30   float64 `json:"1_30_days"`
	Days31To60  float64 `json:"31_60_days"`
	Days61To90  float64 `json:"61_90_days"`
	Over90Days  float64 `json:"over_90_days"`
	Total       float64 `json:"total"`
}

// CustomerAging represents aging for a single customer
type CustomerAging struct {
	CustomerID   uuid.UUID `json:"customer_id"`
	CustomerName string    `json:"customer_name"`
	Current      float64   `json:"current"`
	Days1To30    float64   `json:"1_30_days"`
	Days31To60   float64   `json:"31_60_days"`
	Days61To90   float64   `json:"61_90_days"`
	Over90Days   float64   `json:"over_90_days"`
	Total        float64   `json:"total"`
}

// CashFlowReport represents cash flow report
type CashFlowReport struct {
	Period             ReportPeriod `json:"period"`
	OpeningBalance     float64      `json:"opening_balance"`
	OperatingActivities CashFlowSection `json:"operating_activities"`
	InvestingActivities CashFlowSection `json:"investing_activities"`
	FinancingActivities CashFlowSection `json:"financing_activities"`
	NetCashFlow        float64      `json:"net_cash_flow"`
	ClosingBalance     float64      `json:"closing_balance"`
}

// CashFlowSection represents a section in cash flow
type CashFlowSection struct {
	Inflow  float64 `json:"inflow"`
	Outflow float64 `json:"outflow"`
	Net     float64 `json:"net"`
}
